Test and document keycloak-client requires jdk-11+ runtime

rmartinc opened the pull request:
Closes #135

In order to test with jdk-11 we need to avoid usage of any quarkus or testcontainers because they are jdk-17 based. Ideas:

1. The quarkus dependency for testing is removed and junit5 and log4j-slf4j2 (for the logging) are added directly in the framework sub-project. Previously the quarkus dependency added all this. As a result of this some other test classes needs a small refactor to eliminate classes that were added by quarkus and were used because they were there. Some minor tweaks.
2. As testcontainer for keycloak needs jdk-17 the `remote` lifecycle is used to run the tests for jdk-11. The container is started directly via `podman`. Besides a new `run-tests` workflow is added to be called from the different CIs that need it.
3. Documentation modified that mention in the three libraries that jdk-11 or higher is required at runtime. The release version was also moved to 11 in the main `pom.xml`.

If testcontainers accepts to downgrade java required version to jdk-11 we can back to use it for the jdk-11 workflow.

rmartinc commented:
One more thing, just `nightly` keycloak is used in the jdk-11 tests, but we can add anything we want (right now for normal 21 testing 24, 26, 26.1 and nightly are tested).

jonkoops commented on .github/workflows/run-tests.yml:
```diff
@@ -0,0 +1,76 @@
+    timeout-minutes: 30
+    strategy:
+      matrix:
+        keycloak_server_version: [ "nightly" ]
```
You mentioned in your comment only the nightly is tested for Java 11, is there a particular reason? Or is it just to prevent a massive test execution on CI?

rmartinc replied:
No, no, we can test anything, this was just to setup something. We can do whatever you prefer. But I don't know if we want to test all again.

mposolda replied:
+1 to test just `nightly` with Java 11. IMO sufficient as a compromise, so we have a test coverage for Java 11, but at the same time not need to test whole matrix with Java 11 to save some resources etc.

jonkoops commented on .github/workflows/run-tests.yml:
```diff
@@ -0,0 +1,76 @@
+        name: Test setup
+        uses: ./.github/actions/test-setup
+        with:
+          java-version: |
+            11
+            21
```
Multiple versions of Java can be specified here? How does that work? Should this not be exclusively 11?
```suggestion
          java-version: 11
```

rmartinc replied:
We need the version 21 to create the providers for the server in the first command (because those providers are for the keytcloak server and it cannot be compiled using java-11).:

```bash
mvn -B -f testsuite/providers/pom.xml package -DskipTests=true
```

jonkoops replied:
So the `java-version` can be overloaded into multiple versions? TIL

rmartinc replied:
Yes, GHA allows to install several jdks and then there are env vars like `JAVA_HOME_11_X64` available. Check how the maven is started in the worflow for the jdk-11 test. The general `JAVA_HOME` is the last one (21 in this case).

jonkoops commented on testsuite/admin-client-tests/src/test/java/org/keycloak/client/testsuite/AbstractAdminClientTest.java:
```diff
@@ -212,4 +212,12 @@ protected RealmResource testRealm() {
-}
\ No newline at end of file
+
+    protected String randomAlphanumericString(int count) {
```
Any reason not to use `UUID.randomUUID()` here instead?

rmartinc replied:
This method is used for testing, previously it was `RandomStringUtils.random` using the `commons-lang` dependency obtained from quarkus. It is used to obtain some long string for attributes. I just created a helper in the abstract class to substitute it. The `randomUUID` is of a fixed length. We can do anything here, I preferred to not include more dependencies in the CI. But if you prefer something else just let me know.

jonkoops reviewed (commented):
Changes largely look good, just have a couple of questions.

jonkoops reviewed (dismissed):
LGTM, thanks for the explainer

stianst reviewed (changes requested):
OpenJDK 8 is still supported for quite some time, and we can't just drop support for it without wider discussions first.

jonkoops commented:
> OpenJDK 8 is still supported for quite some time, and we can't just drop support for it without wider discussions first.

It has already been dropped. We already depend on dependencies now that have dropped support for it.

rmartinc commented:
@stianst Currently we are depending on resteasy which does not support jdk 8. There are issues with it (see https://github.com/keycloak/keycloak-client/issues/112). So jdk 8 is certainly not supported since the old resteasy/javax dependencies were removed. Moving backwards and support jdk-8 again means re-adding the old resteasy/javax version which is not maintained (EAP 7 still supports a resteasy 3.15.x version, even older to the one used before, but the new releases are only in the RH repo, no maven central any more). I thought this was already discussed and decided the removal of jdk-8. :smile:

stianst commented:
Not supporting JDK 8 for admin client is one thing, but removing support for all client libraries is a whole different thing

rmartinc commented:
> Not supporting JDK 8 for admin client is one thing, but removing support for all client libraries is a whole different thing

Ah! If resteasy is only used in the admin-client that would be possible. Let me check that. Although the other authz and policy enforcer are even less used. If this works only the admin-client would require the Java 11 or higher.

rmartinc commented:
OK, I have been investigating this a bit more. There are several dependencies that requires jdk 11. Summary:

* resteasy (6.2.9.Final): It is jdk 11 minimum and it's only used by the admin-client. The problem is that the admin client is used in all the testing so we are limited to test using jdk-11 because of this point.
* jboss-logging (3.6.0.Final): It is jdk 11 minimum. And this one is used everywhere. We need to downgrade to 3.4.x (EAP 7.x) and not use the same version quarkus is using inside keycloak project.
* jakarta-activation (2.1.3): It's jakarta ee 10 so it's java 11 minimum. This is only used by the `org.keycloak.common.util.MimeTypeUtil` class included in the `client-common-synced`. I don't know why we are including this class, probably this one can be just removed from the sub-project (not synced from keycloak).
* In the policy-enforcer there are some optional dependencies to jakarta.servlet-api and wildfly-elytron but they are optional. They are also jdk-11 but they are optional so these ones are OK and only thought to be used when used with elytron.

So the summary is that right now the three clients (admin, authz and policy-enforcer) require jdk-11. The `jboss-logging` dep is the main cause and as soon as you reach a class that print some debug mesage you get the following exception:

```
java.lang.UnsupportedClassVersionError: org/jboss/logging/Logger has been compiled by a more recent version of the Java Runtime (class file version 55.0), this version of the Java Runtime only recognizes class file versions up to 52.0
	at java.lang.ClassLoader.defineClass1(Native Method)
	at java.lang.ClassLoader.defineClass(ClassLoader.java:756)
	at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142)
	at java.net.URLClassLoader.defineClass(URLClassLoader.java:473)
	at java.net.URLClassLoader.access$100(URLClassLoader.java:74)
	at java.net.URLClassLoader$1.run(URLClassLoader.java:369)
	at java.net.URLClassLoader$1.run(URLClassLoader.java:363)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.net.URLClassLoader.findClass(URLClassLoader.java:362)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:418)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:352)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:351)
	at org.keycloak.authorization.client.AuthzClient.<clinit>(AuthzClient.java:48)
        ...
```

What do you want to do with this? My feeling is that nobody right now is using the client libs with jdk 8 (you need to manually remove jars and add previous versions). Used inside old version of wildfly (EAP 7.4 times) the libraries can work because the deps are provided by the container. I will ask for a quick meeting tomorrow in our core-clients call because talking is going to be much faster.

rmartinc commented:
I have updated this to maintain jdk 8 in source/target/release. Testing is done using jdk-11 (because with 8 is impossible as the admin-client requires 11). Removed the `jakarta-activation` dependency that was only needed for a class in common (that class was also removed from the repo en the synchronization). The only thing that we should consider is the `jboss-logging` version, right now the version depends on jdk-11. I have removed also the documentation changes until we have a final decision about this.

mposolda commented on testsuite/framework/src/main/java/org/keycloak/testsuite/util/AdminClientUtil.java:
```diff
@@ -109,11 +108,21 @@ public static ResteasyClient createResteasyClient(Boolean followRedirects) {
 
-    private static SSLContext buildSslContext() {
+    public static SSLContext buildSslContext() {
+        return TestConstants.IS_LISFECYCLE_REMOTE
```
IMHO something like using `TestRegistry.INSTANCE.getOrCreateProvider(KeycloakServerProvider.class).createSSLContext()` is slightly better way? For example in case that more implementations would be available, we may not need to update this line. Also no need to create new instance if existing server instance can be re-used etc.

rmartinc replied:
Yep, much better, I just did the first thing I thought and I didn't check there was this `getOrCreateProvider` method. Done!

mposolda reviewed (commented):
@rmartinc @jonkoops Thanks for all the work on this PR and for the review!

Discussed with @stianst about this. We agreed that we can add to the documentation that for the admin-client, we require Java 11 at least.

For authz-client and policy-enforcer, we won't document any Java version and we won't test with Java 8. We can test just with Java 11 as done in this PR. We can see if people report issues or something. (It is a chance that there won't be issues reported as people using EAP7 and jboss-logging 3.4 at runtime will have everything working as expected). Does it look ok to you as well?

So just the doc update for admin-client should be OK IMO? Also added few minor comments inline.

rmartinc commented on testsuite/framework/src/main/java/org/keycloak/client/testsuite/TestConstants.java:
```diff
@@ -6,8 +6,14 @@
+    public static final String PROPERTY_KEYCLOAK_REMOTE_TRUSTSTORE = "keycloak.remote.trustStore";
+    public static final String PROPERTY_KEYCLOAK_REMOTE_TRUSTSTORE_PASSWORD = "keycloak.remote.trustStorePassword";
+
+    public static final boolean IS_LISFECYCLE_REMOTE = KEYCLOAK_LIFECYCLE_REMOTE.equalsIgnoreCase(System.getProperty(PROPERTY_KEYCLOAK_LIFECYCLE));
```
Uups! Yes, fixed!

jonkoops reviewed (approved):
LGTM, no objections here.
